backupccl: Backup Table Statistics

pbardea opened the pull request:
Table statistics for the tables that are being backed up are inserted
into the BackupDescriptor. Upon restoration, the table statistics
are extracted, and re-written to point to the new table IDs and the
relevant statistics are re-inserted into the system.table_statistics
table with an updated created_at date.

In the case of incremental backups and AOST restores, the latest stat
from the relevant incremental backup is restored. However, only one
version of the statistics are currently stored per backup. This
version of the statistics is restored, and a table stats refresh is
triggered.

Important changes:
- The tableStatsitic struct was migrated to use a protobuf in order
to facilitate backwards compatiblity. This ensures that we can
restore backups that may not have the exact same format for table
statistics.

- A Statistics message was added to the BackupDescriptor to store
the relevant table stats as well as any potential future meta-data.

Release note (enterprise change): Backup table statistics for the
tables which are being backed up.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/39644)
<!-- Reviewable:end -->

pbardea committed:
backupccl: Backup Table Statistics

Table statistics for the tables that are being backed up are inserted
into the BackupDescriptor. Upon restoration, the table statistics
are extracted, and re-written to point to the new table IDs and the
relevant statistics are re-inserted into the system.table_statistics
table with an updated created_at date.

In the case of incremental backups and AOST restores, the latest stat
from the relevant incremental backup is restored. However, only one
version of the statistics are currently stored per backup. This
version of the statistics is restored, and a table stats refresh is
triggered.

Important changes:
- The tableStatsitic struct was migrated to use a protobuf in order
to facilitate backwards compatiblity. This ensures that we can
restore backups that may not have the exact same format for table
statistics.

- A Statistics message was added to the BackupDescriptor to store
the relevant table stats as well as any potential future meta-data.

Release note (enterprise change): Backup table statistics for the
tables which are being backed up.

dt reviewed (approved):
> Reviewed 7 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39644)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @pbardea, and @rytaft)

___
*[pkg/ccl/backupccl/restore.go, line 1218 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmMZ2DY3Ta9cuQ_I6fz:-LmMZ2DY3Ta9cuQ_I6g-:b6wmmuk) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/restore.go#L1218)):*
> ```Go
> 
> 	// Update the ID references in the table statistics.
> 	backedUpStatistics := latestBackupDesc.Statistics
> ```

any reason we need to rewrite all the stats?
___
*[pkg/ccl/backupccl/restore.go, line 1687 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmMdQ5E69RVc4lZIJqY:-LmMdQ5E69RVc4lZIJqZ:b-1ze2ux) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/restore.go#L1687)):*
> ```Go
> 		p.ExecCfg().Settings,
> 		backupDescs,
> 		latestBackupDesc,
> ```

2¢: the mutation of the argument is a bit subtle / easy to miss or break if you introduced a copy. I'm fine with it but if you felt like it we return the rewritten stats, or could just rewrite them below instead.


<!-- Sent from Reviewable.io -->

dt reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39644)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @pbardea, and @rytaft)

___
*[pkg/ccl/backupccl/backup_test.go, line 3065 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmMeJDU9FCo6tN7BgNz:-LmMeJDU9FCo6tN7BgO-:bbidgqq) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/backup_test.go#L3065)):*
> ```Go
> 
> 	sqlDB.Exec(t, `CREATE STATISTICS bank_stats FROM data.bank`)
> 	sqlDB.Exec(t, `BACKUP data.bank TO $1 WITH revision_history`, localFoo)
> ```

can you add a test that backs up more tables than it restores?


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39644)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pbardea and @rytaft)

___
*[pkg/ccl/backupccl/restore.go, line 1224 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmMfbjy93BKFj89SH4P:-LmMfbjy93BKFj89SH4Q:bdaxske) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/restore.go#L1224)):*
> ```Go
> 			stats[i] = &latestBackupDesc.Statistics.TableStatistics[i]
> 		}
> 		if err := remapTableIDInStatistics(stats, tableRewrites); err != nil {
> ```

what happens if the backup contains more tables than we have allocated rewrites for (i.e. if we restore one table of a multi-table backup)?
___
*[pkg/ccl/backupccl/restore.go, line 1687 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmMdQ5E69RVc4lZIJqY:-LmMfMiU9tpITzxW04km:b-pjv6ed) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/restore.go#L1687)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

2¢: the mutation of the argument is a bit subtle / easy to miss or break if you introduced a copy. I'm fine with it but if you felt like it we return the rewritten stats, or could just rewrite them below instead.
</blockquote></details>

edit: (if you did choose to do the rewrite below here, you could also combine it with getRelevant to only rewrite the need ones).


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 9 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39644)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pbardea)

___
*[pkg/ccl/backupccl/backup.proto, line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLRb1b1D9FnxeBvEze:-LmLRb1b1D9FnxeBvEzf:b9c3k2s) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/backup.proto#L104)):*
> ```protobuf
>   repeated string partition_descriptor_filenames = 19;
>   repeated string locality_kvs = 20 [(gogoproto.customname) = "LocalityKVs"];
>   message Statistics {
> ```

why do you need a separate message for this?
___
*[pkg/ccl/backupccl/backup_test.go, line 3057 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLak-u2OTaFP_6uJlW:-LmLak-vAHruPtVkdpk1:bfwy1kf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/backup_test.go#L3057)):*
> ```Go
> }
> 
> func TestBackupCreatedStats(t *testing.T) {
> ```

You might want to disable automatic stats for both of these tests to avoid flakiness
___
*[pkg/ccl/backupccl/backup_test.go, line 3070 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLaMvi6yu2DRhUT7C4:-LmLaMvi6yu2DRhUT7C5:b-cvaami) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/backup_test.go#L3070)):*
> ```Go
> 		localFoo, "data 2")
> 
> 	sqlDB.CheckQueryResultsRetry(t,
> ```

Do you need to use retry here? I think the above SQL statements are synchronous.
___
*[pkg/ccl/backupccl/backup_test.go, line 3105 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLaeg85p5bJVQ0VZed:-LmLaeg85p5bJVQ0VZee:b-1770dg) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/backup_test.go#L3105)):*
> ```Go
> 		localFoo, incremental1Foo, incremental2Foo, beforeTs, "data 2"))
> 
> 	sqlDB.CheckQueryResultsRetry(t,
> ```

Ditto -- do you need to use retry?
___
*[pkg/ccl/backupccl/backup_test.go, line 3111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLbKtz0qFW6YWkovJw:-LmLbKtz0qFW6YWkovJx:b-swjvis) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/backup_test.go#L3111)):*
> ```Go
> 			{"bank_stats", "{id}", "3", "3", "0"},
> 			{"bank_stats", "{balance}", "3", "3", "0"},
> 			{"bank_stats", "{payload}", "3", "2", "2"},
> ```

Can you add some comments somewhere about why these are the expected results?
___
*[pkg/ccl/backupccl/restore.go, line 1655 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLd1fo3b4ugFWOIr-c:-LmLd1fo3b4ugFWOIr-d:b-8awxno) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/restore.go#L1655)):*
> ```Go
> 
> 	relevantTableStatistics := make([]*stats.TableStatistic, 0, len(backup.Statistics.TableStatistics))
> 	relevantTables := make(map[sqlbase.ID]bool)
> ```

I'd use `FastIntSet` here, and then you can just check if `relevantTables.Contains(stat.TableID)`
___
*[pkg/ccl/backupccl/restore.go, line 1731 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLdhvg9IPIR6Z7X2ja:-LmLdhvg9IPIR6Z7X2jb:b-wby1bs) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/restore.go#L1731)):*
> ```Go
> 
> 	if err := stats.ReinsertStat(ctx, r.exec, txn, r.latestStats); err != nil {
> 		return errors.Wrapf(err, "count not reinsert table statistics")
> ```

count -> could
___
*[pkg/sql/stats/new_stat.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmPGVHV8Lyd1dTEIMB5:-LmPGVHV8Lyd1dTEIMB6:b-tyx46f) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/sql/stats/new_stat.go#L30)):*
> ```Go
> // Currently used when restoring a backup with table
> // statistics.
> func ReinsertStat(
> ```

[nit] it doesn't really matter that the stats here existed before or not.... I'd rename this function to `InsertNewStats` (i.e. the only difference between this and the below function is that this one is plural)
___
*[pkg/sql/stats/table_stat.proto, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLNL1eDGl6TTVK5Yo5:-LmLNL1f8roNBWoj97ne:bb6ht2j) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/sql/stats/table_stat.proto#L1)):*
> ```protobuf
> // Copyright 2017 The Cockroach Authors.
> ```

2019


<!-- Sent from Reviewable.io -->

pbardea reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39644)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @rytaft)

___
*[pkg/ccl/backupccl/backup.proto, line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLRb1b1D9FnxeBvEze:-LmQRiGS9aR-ai4CkE4z:b-pt3xrw) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/backup.proto#L104)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

why do you need a separate message for this?
</blockquote></details>

The original intention was that if we were going to store any other information about the statistics in the future. Such as metadata needed for backing up AOST and such we would could group those fields in this message. If it's a top level entry for now it would be hard to change later if we want to include such data.
___
*[pkg/ccl/backupccl/backup_test.go, line 3057 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLak-u2OTaFP_6uJlW:-LmPvvAF-3oWU6fLXfQr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/backup_test.go#L3057)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

You might want to disable automatic stats for both of these tests to avoid flakiness
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/backup_test.go, line 3065 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmMeJDU9FCo6tN7BgNz:-LmQRhDN2ph7DIl7iyfy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/backup_test.go#L3065)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

can you add a test that backs up more tables than it restores?
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/backup_test.go, line 3070 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLaMvi6yu2DRhUT7C4:-LmPxNAeAw9s_jtDUlL6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/backup_test.go#L3070)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Do you need to use retry here? I think the above SQL statements are synchronous.
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/backup_test.go, line 3105 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLaeg85p5bJVQ0VZed:-LmPxNl4CTZPzhKgYOCY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/backup_test.go#L3105)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Ditto -- do you need to use retry?
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/backup_test.go, line 3111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLbKtz0qFW6YWkovJw:-LmPxy8-DWud6u4s-Dmf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/backup_test.go#L3111)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Can you add some comments somewhere about why these are the expected results?
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/restore.go, line 1218 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmMZ2DY3Ta9cuQ_I6fz:-LmQRdXF9LT_F_W597Yq:b-2wzuxl) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/restore.go#L1218)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

any reason we need to rewrite all the stats?
</blockquote></details>

Ended up replacing this.
___
*[pkg/ccl/backupccl/restore.go, line 1224 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmMfbjy93BKFj89SH4P:-LmQRJPJDy7ywUDw6xOO:bemoep6) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/restore.go#L1224)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

what happens if the backup contains more tables than we have allocated rewrites for (i.e. if we restore one table of a multi-table backup)?
</blockquote></details>

Done. Added a test and moved the entire process to after the `restore` call and 
___
*[pkg/ccl/backupccl/restore.go, line 1655 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLd1fo3b4ugFWOIr-c:-LmQRCeP5Mhrn58Vbx9S:bqrrlm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/restore.go#L1655)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I'd use `FastIntSet` here, and then you can just check if `relevantTables.Contains(stat.TableID)`
</blockquote></details>

I ended up replacing this with just doing the table rewriter lookup.
___
*[pkg/ccl/backupccl/restore.go, line 1687 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmMdQ5E69RVc4lZIJqY:-LmQTjs03YqNM_JhdL9Q:bbzlxq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/restore.go#L1687)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

edit: (if you did choose to do the rewrite below here, you could also combine it with getRelevant to only rewrite the need ones).
</blockquote></details>

Done. Changed it so that all of the stats stuff is in one separate function call that doesn't mutate the backup descriptor and returns a copy.
___
*[pkg/ccl/backupccl/restore.go, line 1731 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLdhvg9IPIR6Z7X2ja:-LmQR8L49BPM2iGa7t99:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/restore.go#L1731)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

count -> could
</blockquote></details>

Done.
___
*[pkg/sql/stats/new_stat.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmPGVHV8Lyd1dTEIMB5:-LmQR9nyD9eCgmi1vOfr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/sql/stats/new_stat.go#L30)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] it doesn't really matter that the stats here existed before or not.... I'd rename this function to `InsertNewStats` (i.e. the only difference between this and the below function is that this one is plural)
</blockquote></details>

Done.
___
*[pkg/sql/stats/table_stat.proto, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLNL1eDGl6TTVK5Yo5:-LmQRAzi46-POS_EZ32z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/sql/stats/table_stat.proto#L1)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

2019
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dt reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39644)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @pbardea, and @rytaft)

___
*[pkg/ccl/backupccl/restore.go, line 1163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmR6e5O6VOIDAswTxG3:-LmR6e5O6VOIDAswTxG4:b-1o5nww) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbedb68dc9c12a914b3f52be262ea54f48bc8e52/pkg/ccl/backupccl/restore.go#L1163)):*
> ```Go
> 	settings *cluster.Settings,
> 	backupDescs []BackupDescriptor,
> 	latestBackupDesc BackupDescriptor,
> ```

is this new argument still used after moving the stats stuff?
___
*[pkg/ccl/backupccl/restore.go, line 1207 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmR6n1B82VEaG2vDSc5:-LmR6n1B82VEaG2vDSc6:b-monr71) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbedb68dc9c12a914b3f52be262ea54f48bc8e52/pkg/ccl/backupccl/restore.go#L1207)):*
> ```Go
> 	spans := spansForAllTableIndexes(tables, nil)
> 
> 	var err error
> ```

is this leftover?


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39644)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @pbardea, and @rytaft)

___
*[pkg/ccl/backupccl/restore.go, line 1625 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmR7Gn-3omR0CRU39dq:-LmR7Gn-3omR0CRU39dr:bxl2wmw) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbedb68dc9c12a914b3f52be262ea54f48bc8e52/pkg/ccl/backupccl/restore.go#L1625)):*
> ```Go
> 	backup BackupDescriptor, tableRewrites TableRewriteMap,
> ) []*stats.TableStatistic {
> 	relevantTableStatistics := make([]*stats.TableStatistic, 0)
> ```

minor nit: if you wanted to, you could add `, len(tableRewrites)` here to avoid growing re-allocs.


<!-- Sent from Reviewable.io -->

pbardea reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39644)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @pbardea, and @rytaft)

___
*[pkg/ccl/backupccl/restore.go, line 1625 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmR7Gn-3omR0CRU39dq:-LmR8E2N46d4eobGt-rJ:b-jsv5jg) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbedb68dc9c12a914b3f52be262ea54f48bc8e52/pkg/ccl/backupccl/restore.go#L1625)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

minor nit: if you wanted to, you could add `, len(tableRewrites)` here to avoid growing re-allocs.
</blockquote></details>

There may be more than one stat column per table and not all tables may have stats, so it'd likely have to grow the capacity further. And specifying it as the number of tables may be confusing.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 7 of 8 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39644)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @pbardea)

___
*[pkg/ccl/backupccl/backup.go, line 1019 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-Lmdk7ouC8vN944QF4AP:-Lmdk7ov-jv6pWjkPeGZ:b-qst278) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/ccl/backupccl/backup.go#L1019)):*
> ```Go
> 				}
> 				for _, tableStatistic := range tableStatisticsAcc {
> 					tableStatistics = append(tableStatistics, tableStatistic)
> ```

Instead of a for loop, you can do `tableStatistics = append(tableStatistics, tableStatisticsAcc...)`
___
*[pkg/ccl/backupccl/backup.proto, line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLRb1b1D9FnxeBvEze:-LmdkXC902l70WndManJ:b6dtvdd) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/backup.proto#L104)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

The original intention was that if we were going to store any other information about the statistics in the future. Such as metadata needed for backing up AOST and such we would could group those fields in this message. If it's a top level entry for now it would be hard to change later if we want to include such data.
</blockquote></details>

Why would AOST metadata be stored in the Statistics message?  

I guess the question is whether there is actually metadata that would apply for all statistics. It's possible that each statistic was originally created with a different AOST value.
___
*[pkg/ccl/backupccl/backup_test.go, line 3064 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-Lmdm7oX5J3E5g3Lr0Yf:-Lmdm7oX5J3E5g3Lr0Yg:b-g2b1dm) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/ccl/backupccl/backup_test.go#L3064)):*
> ```Go
> 	defer cleanupFn()
> 
> 	sqlDB.Exec(t, `SET CLUSTER SETTING sql.stats.automatic_collection.enabled=true`)
> ```

This should be `=false` to disable automatic stats
___
*[pkg/ccl/backupccl/backup_test.go, line 3097 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmdmBnl8KN0hgtDfLOh:-LmdmBnl8KN0hgtDfLOi:bhgm7m7) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/ccl/backupccl/backup_test.go#L3097)):*
> ```Go
> 	defer cleanupFn()
> 
> 	sqlDB.Exec(t, `SET CLUSTER SETTING sql.stats.automatic_collection.enabled=true`)
> ```

ditto -- should be `=false`
___
*[pkg/ccl/backupccl/backup_test.go, line 3130 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmdmRwbCa-V07xAZrcz:-LmdmRwbCa-V07xAZrd-:b-g9ywjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/ccl/backupccl/backup_test.go#L3130)):*
> ```Go
> 	var beforeTs string
> 
> 	sqlDB.Exec(t, `SET CLUSTER SETTING sql.stats.automatic_collection.enabled=true`)
> ```

`=false`
___
*[pkg/ccl/backupccl/restore.go, line 1625 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmR7Gn-3omR0CRU39dq:-Lmdno345hvvoxwhbhn9:bbmm7mp) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbedb68dc9c12a914b3f52be262ea54f48bc8e52/pkg/ccl/backupccl/restore.go#L1625)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

There may be more than one stat column per table and not all tables may have stats, so it'd likely have to grow the capacity further. And specifying it as the number of tables may be confusing.
</blockquote></details>

If you move it below the check for nil, you could allocate capacity for `len(backup.Statistics.TableStatistics)` like you did before...
___
*[pkg/ccl/backupccl/restore.go, line 1618 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmdokbRD7X-CiNc1Wli:-LmdokbRD7X-CiNc1Wlj:b-69n31h) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/ccl/backupccl/restore.go#L1618)):*
> ```Go
> // remapRelevantStatistics extracts the stats stored in the backup which correspond
> // to any of the given tables. Note that the given tables must have the old table
> // IDs.
> ```

[nit] This comment is a bit confusing now that you've changed to using `tableRewrites`
___
*[pkg/sql/stats/new_stat.go, line 25 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-Lmdp9Ap3dlBf79E0sIp:-Lmdp9Ap3dlBf79E0sIq:bzve6f) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/sql/stats/new_stat.go#L25)):*
> ```Go
> )
> 
> // InsertNewStats is used to restore an old statistic
> ```

update comment to remove reference to "old statistic"


<!-- Sent from Reviewable.io -->

pbardea reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39644)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @rytaft)

___
*[pkg/ccl/backupccl/backup.go, line 1019 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-Lmdk7ouC8vN944QF4AP:-Lme-Ve6BmOhQX3hOqYz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/ccl/backupccl/backup.go#L1019)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Instead of a for loop, you can do `tableStatistics = append(tableStatistics, tableStatisticsAcc...)`
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/backup.proto, line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLRb1b1D9FnxeBvEze:-Lme-WY-5U1kaoR12WLZ:b-ys4egn) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/backup.proto#L104)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Why would AOST metadata be stored in the Statistics message?  

I guess the question is whether there is actually metadata that would apply for all statistics. It's possible that each statistic was originally created with a different AOST value.
</blockquote></details>

It looks like since stats are not updated and a new one is created when the stats change that there likely won't be any metadata that we need to store. I've changed this property to be at the top level of the BackupDescriptor.

The only other use case would be if there were any other statistics we would eventually want to store, but it seems that relational expression statistics are actually stored in the same format.
___
*[pkg/ccl/backupccl/backup_test.go, line 3064 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-Lmdm7oX5J3E5g3Lr0Yf:-Lme0ByO-VerGVH9Jxi9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/ccl/backupccl/backup_test.go#L3064)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

This should be `=false` to disable automatic stats
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/backup_test.go, line 3097 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmdmBnl8KN0hgtDfLOh:-Lme0CMS7OXfYTEygKT-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/ccl/backupccl/backup_test.go#L3097)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

ditto -- should be `=false`
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/backup_test.go, line 3130 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmdmRwbCa-V07xAZrcz:-Lme0CiI6wV3z_hz9XMO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/ccl/backupccl/backup_test.go#L3130)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

`=false`
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/restore.go, line 1163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmR6e5O6VOIDAswTxG3:-Lme0DvJ0y9-8OJYhxBh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbedb68dc9c12a914b3f52be262ea54f48bc8e52/pkg/ccl/backupccl/restore.go#L1163)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

is this new argument still used after moving the stats stuff?
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/restore.go, line 1207 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmR6n1B82VEaG2vDSc5:-Lme0Ecr3szJzSu20QKB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbedb68dc9c12a914b3f52be262ea54f48bc8e52/pkg/ccl/backupccl/restore.go#L1207)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

is this leftover?
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/restore.go, line 1618 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmdokbRD7X-CiNc1Wli:-Lme0FNa7ffjl3wfYkJp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/ccl/backupccl/restore.go#L1618)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] This comment is a bit confusing now that you've changed to using `tableRewrites`
</blockquote></details>

Done.
___
*[pkg/sql/stats/new_stat.go, line 25 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-Lmdp9Ap3dlBf79E0sIp:-Lme0I9r3nncZcW621tS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/sql/stats/new_stat.go#L25)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

update comment to remove reference to "old statistic"
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

pbardea reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39644)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @rytaft)

___
*[pkg/ccl/backupccl/backup.go, line 1019 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-Lmdk7ouC8vN944QF4AP:-Lme-Ve6BmOhQX3hOqYz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/ccl/backupccl/backup.go#L1019)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Instead of a for loop, you can do `tableStatistics = append(tableStatistics, tableStatisticsAcc...)`
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/backup.proto, line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmLRb1b1D9FnxeBvEze:-Lme-WY-5U1kaoR12WLZ:b-ys4egn) ([raw file](https://github.com/cockroachdb/cockroach/blob/cce30d885dbae0b2a2b358d3f8d51a0c248b32ee/pkg/ccl/backupccl/backup.proto#L104)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Why would AOST metadata be stored in the Statistics message?  

I guess the question is whether there is actually metadata that would apply for all statistics. It's possible that each statistic was originally created with a different AOST value.
</blockquote></details>

It looks like since stats are not updated and a new one is created when the stats change that there likely won't be any metadata that we need to store. I've changed this property to be at the top level of the BackupDescriptor.

The only other use case would be if there were any other statistics we would eventually want to store, but it seems that relational expression statistics are actually stored in the same format.
___
*[pkg/ccl/backupccl/backup_test.go, line 3064 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-Lmdm7oX5J3E5g3Lr0Yf:-Lme0ByO-VerGVH9Jxi9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/ccl/backupccl/backup_test.go#L3064)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

This should be `=false` to disable automatic stats
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/backup_test.go, line 3097 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmdmBnl8KN0hgtDfLOh:-Lme0CMS7OXfYTEygKT-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/ccl/backupccl/backup_test.go#L3097)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

ditto -- should be `=false`
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/backup_test.go, line 3130 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmdmRwbCa-V07xAZrcz:-Lme0CiI6wV3z_hz9XMO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/ccl/backupccl/backup_test.go#L3130)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

`=false`
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/restore.go, line 1163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmR6e5O6VOIDAswTxG3:-Lme0DvJ0y9-8OJYhxBh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbedb68dc9c12a914b3f52be262ea54f48bc8e52/pkg/ccl/backupccl/restore.go#L1163)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

is this new argument still used after moving the stats stuff?
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/restore.go, line 1207 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmR6n1B82VEaG2vDSc5:-Lme0Ecr3szJzSu20QKB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbedb68dc9c12a914b3f52be262ea54f48bc8e52/pkg/ccl/backupccl/restore.go#L1207)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

is this leftover?
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/restore.go, line 1618 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-LmdokbRD7X-CiNc1Wli:-Lme0FNa7ffjl3wfYkJp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/ccl/backupccl/restore.go#L1618)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] This comment is a bit confusing now that you've changed to using `tableRewrites`
</blockquote></details>

Done.
___
*[pkg/sql/stats/new_stat.go, line 25 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-Lmdp9Ap3dlBf79E0sIp:-Lme0I9r3nncZcW621tS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fed8e36cc29457e287761bd05852251c09dc0b02/pkg/sql/stats/new_stat.go#L25)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

update comment to remove reference to "old statistic"
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dt reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39644)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)



<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks!

> Reviewed 2 of 6 files at r4, 4 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39644#-:-LmeFdsj62ZXv1C7sqkF:bww03zd)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

pbardea commented:
Thank you both for the review!

bors r+
